File: src/advanced-ds/rbtree.ml
open Batteries

type color = Red | Black
type 'a tree = Empty | Tree of color * 'a tree * 'a * 'a tree

let empty = Empty

let rec mem x = function
  | Empty             -> false
  | Tree (_, a, y, b) ->
    if x < y then mem x a
    else if x = y then true
    else mem x b

let balance color a x b =
  match color, a, x, b with
    | Black, Tree (Red, Tree (Red, a, x, b), y, c), z, d ->
      Tree (Red, Tree (Black, a, x, b), y, Tree (Black, c, z, d))
    | Black, Tree (Red, a, x, Tree (Red, b, y, c)), z, d ->
      Tree (Red, Tree (Black, a, x, b), y, Tree (Black, c, z, d))
    | Black, a, x, Tree (Red, Tree (Red, b, y, c), z, d) ->
      Tree (Red, Tree (Black, a, x, b), y, Tree (Black, c, z, d))
    | Black, a, x, Tree (Red, b, y, Tree (Red, c, z, d)) ->
      Tree (Red, Tree (Black, a, x, b), y, Tree (Black, c, z, d))
    | _ -> Tree (color, a, x, b)

let insert x tree =
  let make_black = function
    | Tree (_, a, y, b) -> Tree (Black, a, y, b)
    | Empty             -> raise (Failure "Empty tree passed") in
  let rec insert' = function
    | Empty -> Tree (Red, Empty, x, Empty)
    | Tree (color, a, y, b) ->
      if      x < y then balance color (insert' a) y b
      else if x = y then Tree (color, a, y, b)
      else               balance color a y (insert' b)
  in
  make_black (insert' tree)

File: src/advanced-ds/rbtree.hs
module Main
       where

data Color = R
           | B
           deriving Show

data Tree e = E
            | T Color (Tree e) e (Tree e)
            deriving Show

empty :: Tree e
empty = E

member :: Ord e => e -> Tree e -> Bool
member x E = False
member x (T _ a y b) | x <  y = member x a
                     | x == y = True
                     | x >  y = member x b

insert :: Ord e => e -> Tree e -> Tree e
insert x t = makeBlack (insert' t)
  where insert' E = T R E x E
        insert' (T color a y b) | x <  y = balance color (insert' a) y b
                                | x == y = T color a y b
                                | x >  y = balance color a y (insert' b)
        makeBlack (T _ a x b) = T B a x b

balance :: Color -> Tree e -> e -> Tree e -> Tree e
balance B (T R (T R a x b) y c) z d = T R (T B a x b) y (T B c z d)
balance B (T R a x (T R b y c)) z d = T R (T B a x b) y (T B c z d)
balance B a x (T R b y (T R c z d)) = T R (T B a x b) y (T B c z d)
balance B a x (T R (T R b y c) z d) = T R (T B a x b) y (T B c z d)
balance color a x b = T color a x b

fromList :: Ord e => [e] -> Tree e
fromList = foldl (flip insert) E

main :: IO ()
main = do
  putStrLn $ show $ fromList [9, 8, 7, 4, 5, 6, 3, 2, 1, 0]
